Ansible task

zmc opened the pull request.

ceph-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ceph.com//job/teuthology-pull-requests/1440/
Tests passed for this pull request.

andrewschoen commented on teuthology/task/ansible.py:
```diff
@@ -0,0 +1,225 @@
+log = logging.getLogger(__name__)
+
+
+class LoggerFile(object):
```
Why do we need this?  Maybe the docstring should explain that?

zmc replied:
Explanation added

andrewschoen replied:
That doesn't explain the why though.  Why can't ansible use the standard logger?

zmc replied:
Well of course ansible itself doesn't know about python loggers. Normally what one would probably do is use ``subprocess.Popen`` to run the process, tell it to pipe stderr and stdout to a buffer, then call ``readlines()`` on the buffer in a loop and log the output. For whatever reason that doesn't work with ``ansible-playbook``; perhaps it writes directly to the terminal instead of using stdout/stderr.

``pexpect`` handles cases like this, and it takes a ``logfile`` argument, so I built a logger with a ``file``-like interface.

andrewschoen commented on teuthology/task/ansible.py:
```diff
@@ -0,0 +1,225 @@
+        repo: /path/to/repo
+        inventory: /path/to/inventory
+        playbook: /path/to/playbook.yml
+        tags: my_tags
```
tags is not documented above.  Does this take a comma separated list of tags?

zmc replied:
Oh, you're right! I'll document it.

zmc replied:
Done. And ``branch``.

andrewschoen commented on teuthology/task/ansible.py:
```diff
@@ -0,0 +1,225 @@
+        self.generated_playbook = False
+
+    def setup(self):
+        super(Ansible, self).setup()
```
This isn't needed, right?

andrewschoen commented on teuthology/task/ansible.py:
```diff
@@ -0,0 +1,225 @@
+        Locate the repo we're using; cloning it from a remote repo if necessary
+        """
+        repo = self.config.get('repo', '.')
+        if repo.startswith('http://') or repo.startswith('https://') or \
```
nitpick: when using long conditions like this I like to either use a variable to hold the condition or create a small, easily tested function.  I find it helps with readability.

andrewschoen commented on teuthology/task/ansible.py:
```diff
@@ -0,0 +1,225 @@
+        if isinstance(playbook, list):
+            # Multiple plays in a list
+            self.playbook = playbook
+        elif isinstance(playbook, str) and (playbook.startswith('http://') or \
```
nitpick: checking again here to see if something is a url.  Maybe a utility method for these types of checks would be worth it?

andrewschoen commented on teuthology/task/ansible.py:
```diff
@@ -0,0 +1,225 @@
+        using an existing file.
+        """
+        for play in self.playbook:
+            play['hosts'] = 'all'
```
I understand why you do this, but a comment might help the next reader of this task.

andrewschoen commented on teuthology/task/ansible.py:
```diff
@@ -0,0 +1,225 @@
+        self.generated_playbook = True
+
+    def begin(self):
+        super(Ansible, self).begin()
```
Not needed.

andrewschoen commented on teuthology/task/ansible.py:
```diff
@@ -0,0 +1,225 @@
+        super(Ansible, self).begin()
+        self.execute_playbook()
+
+    def execute_playbook(self, _logfile=None):
```
I assume _logfile is for test injection purposes.  If so, can we doc that?

zmc replied:
Done

andrewschoen commented on teuthology/task/ansible.py:
```diff
@@ -0,0 +1,225 @@
+        Execute ansible-playbook
+        """
+        environ = os.environ
+        environ['ANSIBLE_SSH_PIPELINING'] = '1'
```
\o/

andrewschoen commented on teuthology/task/ansible.py:
```diff
@@ -0,0 +1,225 @@
+        """
+        Assemble the list of args to be executed
+        """
+        fqdns = [r.name.split('@')[-1] for r in self.cluster.remotes.keys()]
```
Don't you only want the fqdns for what's listed in hosts and not all the remotes that teuthology currently has locked?

zmc replied:
``self.cluster`` has been filtered already! Thanks, base ``Task``!

andrewschoen replied:
Ah, that's right.  We should probably document that behavior in the docstring of Task at some point.

zmc replied:
Done

andrewschoen commented on teuthology/task/ansible.py:
```diff
@@ -0,0 +1,225 @@
+            os.remove(self.inventory)
+        if self.generated_playbook:
+            os.remove(self.playbook_file.name)
+        super(Ansible, self).teardown()
```
Not needed.

andrewschoen commented on teuthology/test/task/test_ansible.py:
```diff
@@ -0,0 +1,306 @@
+            ansible.Ansible,
+            find_repo=DEFAULT,
+            get_playbook=fake_get_playbook,
+            get_inventory=fake_get_inventory,
```
Could just use Mock() here.

andrewschoen commented on teuthology/test/task/test_ansible.py:
```diff
@@ -0,0 +1,306 @@
+        with patch.multiple(
+            ansible.Ansible,
+            find_repo=DEFAULT,
+            get_playbook=fake_get_playbook,
```
Could use Mock(side_effect=Exception) here.

andrewschoen commented on teuthology/test/task/test_ansible.py:
```diff
@@ -0,0 +1,306 @@
+            get_playbook=fake_get_playbook,
+            get_inventory=fake_get_inventory,
+            generate_hosts_file=DEFAULT,
+            generate_playbook=fake_generate_playbook,
```
Could use Mock(side_effect=Exception) here.

andrewschoen commented on teuthology/test/task/test_ansible.py:
```diff
@@ -0,0 +1,306 @@
+        def fake_get_inventory(self):
+            pass
+
+        with patch.multiple(
```
patch.multiple, very cool!

/me didn't know this was a thing

ceph-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ceph.com//job/teuthology-pull-requests/1441/
Tests passed for this pull request.

zmc commented:
Addressed the docs-related comments.

As we discussed, the ``super()`` calls will stay at least for now.

I changed the long ``if`` statement to be more readable, but I don't want to move the content of the ``elif`` as I think that would actually be slightly more confusing.

ceph-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ceph.com//job/teuthology-pull-requests/1442/
Tests passed for this pull request.

ceph-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ceph.com//job/teuthology-pull-requests/1443/
Tests passed for this pull request.

ceph-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ceph.com//job/teuthology-pull-requests/1444/
Tests passed for this pull request.

andrewschoen commented:
This is looking really good.  Once we clean up the commits and maybe do some more docs clarification I think this is good to go. 

Concerning the calls to ``super``, I still feel pretty strongly that the ``setup``, ``begin``, ``end`` and ``teardown`` methods of ``Task`` should be reserved for use by subclasses of ``Task``, but for right now those are noops so I won't block on this PR as we discuss that.

Nice work!

alfredodeza commented on teuthology/task/ansible.py:
```diff
@@ -0,0 +1,239 @@
+        Locate the repo we're using; cloning it from a remote repo if necessary
+        """
+        repo = self.config.get('repo', '.')
+        is_remote = (repo.startswith('http://') or
```
`startswith` can take a tuple of strings to try, instead of doing `or` multiple times:

    is_remote = repo.startswith(('http', 'git', 'https'))

andrewschoen replied:
I didn't know about that!  That would be nice here, +1.

zmc replied:
Didn't know that @alfredodeza, thanks!

alfredodeza commented on teuthology/task/ansible.py:
```diff
@@ -0,0 +1,239 @@
+        if isinstance(playbook, list):
+            # Multiple plays in a list
+            self.playbook = playbook
+        elif isinstance(playbook, str) and (playbook.startswith('http://') or
```
if the check 'http' or 'https' happens in more than one place you could just have a helper

ceph-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.ceph.com//job/teuthology-pull-requests/1445/
Tests passed for this pull request.
